File: DP/1143.longest_common_subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        
        # Make a grid of 0's with len(text2) + 1 columns 
        # and len(text1) + 1 rows.
        dp_grid = [[0] * (len(text2) + 1) for _ in range(len(text1) + 1)]
        
        # Iterate up each column, starting from the last one.
        
        for row in reversed(range(len(text1))):
            for col in reversed(range(len(text2))):
                # If the corresponding characters for this cell are the same...
                if text1[row] == text2[col]:
                    dp_grid[row][col] = 1 + dp_grid[row + 1][col + 1]
                # Otherwise they must be different...
                else:
                    dp_grid[row][col] = max(dp_grid[row + 1][col], dp_grid[row][col + 1])
        
        # The original problem's answer is in dp_grid[0][0]. Return it.
        return dp_grid[0][0]

# Space optimization, with two rows instead of entire grid
class Solution2:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        # The previous and current column starts with all 0's and like 
        # before is 1 more than the length of the first word.
        previous = [0] * (len(text2) + 1)
        current = [0] * (len(text2) + 1)
        # Iterate up each column, starting from the last one.
        for row in reversed(range(len(text1))):
            for col in reversed(range(len(text2))):
                if text1[row] == text2[col]:
                    current[col] = 1 + previous[col + 1]
                else:
                    current[col] = max(previous[col], current[col + 1])
            # The current column becomes the previous one, and vice versa.
            previous, current = current, previous
        
        # The original problem's answer is in previous[0]. Return it.
        return previous[0]

File: C++/Dynamic Programming/1143. Longest Common Subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int m = text1.size(), n = text2.size();
        if (m < n) return longestCommonSubsequence(text2, text1);
        vector<vector<int>> dp(2, vector<int>(n+1, 0));
        int dp_row = 0;
        for (int i = 1; i <= m; i++) {
            dp_row = 1 - dp_row;
            for (int j = 1; j <= n; j++) {
                if (text1[i-1] == text2[j-1]) {
                    dp[dp_row][j] = 1 + dp[1 - dp_row][j-1];
                }
                else {
                    dp[dp_row][j] = max(dp[1 - dp_row][j], dp[dp_row][j-1]);
                }
            }
        }
        return dp[dp_row][n];
    }
};